packages/protocol/contracts/L1/verifiers/SgxVerifier2.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { ECDSAUpgradeable } from "@ozu/utils/cryptography/ECDSAUpgradeable.sol";
+
+import { EssentialContract } from "../../common/EssentialContract.sol";
+import { Proxied } from "../../common/Proxied.sol";
+import { LibBytesUtils } from "../../thirdparty/LibBytesUtils.sol";
+
+import { TaikoData } from "../TaikoData.sol";
+
+import { IVerifier } from "./IVerifier.sol";
+
+/// @title SgxVerifier2
+contract SgxVerifier2 is EssentialContract, IVerifier {
+    using ECDSAUpgradeable for bytes32;
+
+    uint256 public constant INSTANCE_EXPIRY = 180 days;
+
+    mapping(address instance => uint256 registeredAt) public instances;
+
+    uint256[49] private __gap;
+
+    event InstanceRegistered(address indexed instance, uint256 registeredAt);
+
+    event InstanceDeleted(address indexed instance, uint256 deletedAt);
+
+    error SGX_INSTANCE_NOT_FOUND();
+    error SGX_INSTANCE_REGISTERED();
+    error SGX_INVALID_INSTANCE();
+    error SGX_INVALID_PROOF_SIZE();
+
+    /// @notice Initializes the contract with the provided address manager.
+    /// @param _addressManager The address of the address manager contract.
+    function init(address _addressManager) external initializer {
+        EssentialContract._init(_addressManager);
+    }
+
+    function registerInstance(address instance) external onlyOwner {
+        _registerInstance(instance);
+    }
+
+    function updateInstance(
+        address[] calldata newInstances,
+        bytes memory signature
+    )
+        external
+    {
+        address oldInstance = keccak256(
+            abi.encode("REGISTER_SGX_INSTANCE", newInstances)
+        ).recover(signature);
+
+        _deregisterInstance(oldInstance);
+
+        for (uint256 i; i < newInstances.length; ++i) {
+            _registerInstance(newInstances[i]);
+        }
+    }
+
+    /// @inheritdoc IVerifier
+    function verifyProof(
+        uint64,
+        address prover,
+        bool isContesting,
+        TaikoData.BlockEvidence calldata evidence
+    )
+        external
+        onlyFromNamed("taiko")
+    {
+        // Do not run proof verification to contest an existing proof
+        if (isContesting) return;
+
+        // Size is: 87 bytes
+        // 2 bytes + 20 bytes + 65 bytes = 87
+        if (evidence.proof.length <= 20) {
+            revert SGX_INVALID_PROOF_SIZE();
+        }
+
+        address newInstance =
```

Comment:
@adaki2004 One question I have is whether we should make "newInstance" here also a list: `newInstances`, given that the `replaceInstance()` function takes an array.